R4R: Simulate transactions for fee distribution, simulate inflation, and fix a multitude of bugs discovered in the process of doing so

cwgoes opened the pull request:
Closes https://github.com/cosmos/cosmos-sdk/issues/1870
Closes https://github.com/cosmos/cosmos-sdk/issues/2368
Closes #2485 

Add transactions (and maybe invariants?) for the new fee distribution module.

Standard checklist:

- [x] Linked to github-issue with discussion and accepted design OR link to spec that describes this work.
- [x] Wrote tests
- [x] Updated relevant documentation (`docs/`)
- [x] Added entries in `PENDING.md` with issue # 
- [x] rereviewed `Files changed` in the github PR explorer

______

For Admin Use:
- Added appropriate labels to PR (ex. wip, ready-for-review, docs)
- Reviewers Assigned
- Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

rigelrozanski commented:
Thanks for starting this Chris!

cwgoes commented:
I think we need to add some sanity checks:

- For nonexistent delegation reward info on `WithdrawDelegatorReward` (maybe the delegator never delegated to the validator)
- For nonexistent validator reward info on `WithdrawValidatorRewardsAll` (maybe the provided address isn't even a validator)
- For nonexistent delegation reward info on `WithdrawDelegatorRewardsAll` (maybe the address has never delegated)

Also, when I comment out those three messages, supply conservation is broken:

```bash
Running quick Gaia simulation. This may take several minutes...
=== RUN   TestFullGaiaSimulation
Starting SimulateFromSeed with randomness created with seed 42
Starting the simulation from time Sun Jan 31 13:59:47 UTC 26140, unixtime 762734152787
Simulating... block 27/400, operation 200/203.  expected loose tokens to equal total steak held by accounts - pool.LooseTokens: 9738, sum of account tokens: 9737
Too many logs to display, instead writing to simulation_log_2018-10-17 00:43:37.txt
--- FAIL: TestFullGaiaSimulation (2.88s)
    util.go:111: 
FAIL
FAIL	github.com/cosmos/cosmos-sdk/cmd/gaia/app	2.899s
make: *** [Makefile:172: test_sim_gaia_fast] Error 1
```

I also fixed a bug in the simulation where power wasn't set correctly for last-commit, so we may not really have been simulating fee distribution at all before.

rigelrozanski commented:
All good points -> with regards to fee distribution conservation of tokens - this very well could be due to the rounding down of tokens when you withdraw (and you are due a non-integer amount). - one easy way to deal with this in the short term is to just donate any trimmings of tokens to the community tax pool - again this will be extremely small amounts as we're going to be dealing with nano-atoms here

cwgoes commented:
Hmm this passes now but I don't think we're actually calling `SetCollectedFees` anywhere... also we might need to actually charge fees in the simulation.

cwgoes commented:
OK, I've found more problems - after disabling rounding the supply (which the simulation used to do) it fails even without any of the new fee distribution logic at all. I think we might be rounding somewhare in `x/stake` and not dealing with the remainder correctly.

rigelrozanski commented:
> after disabling rounding the supply (which the simulation used to do) it fails even without any of the new fee distribution logic at all. I think we might be rounding somewhare in x/stake and not dealing with the remainder correctly. 

Can you highlight this change in with a commit? 
@cwgoes

cwgoes commented:
> Can you highlight this change in with a commit?

I've fixed that, I think - indeed, we were rounding without accounting for the change - https://github.com/cosmos/cosmos-sdk/pull/2501/commits/42ae2671c2bf6e9da442cba49df6416da1185e01#diff-b905817409b4c0b71e9adffb102682b4R483.

Now debugging fee distribution, which breaks the supply invariant if enabled.

cwgoes commented:
Simulation currently failing after `MsgWithdrawDelegatorRewardsAll` with more account tokens than expected:

```bash
 Invariants broken after operation: TestMsgWithdrawDelegatorRewardsAll: ok true, msg {"type":"cosmos-sdk/MsgWithdrawDelegationRewardsAll","value":{"delegator_addr":"cosmos1smxzx0l76svsqwr2dplfpe2ua4e5xqdttkj0nj"}}
expected loose tokens to equal total steak held by accounts - pool.LooseTokens: 233620000000000, sum of account tokens: 233623011933986
```

cwgoes commented:
OK, *now* this fails with:

```bash
Starting SimulateFromSeed with randomness created with seed 42
Starting the simulation from time Sun Jan 31 13:59:47 UTC 26140, unixtime 762734152787
Simulating... block 34/400, operation 0/122.  Panic with err
 ERROR:
Codespace: 1
Code: 10
Message: "Error{1steak < -2steak}"

goroutine 16 [running]:
runtime/debug.Stack(0xc002ada620, 0x2, 0x2)
	/usr/lib/go/src/runtime/debug/stack.go:24 +0xa7
github.com/cosmos/cosmos-sdk/x/mock/simulation.SimulateFromSeed.func3(0xc000b8c400, 0xc0000d8700, 0xc0001e7170)
	/home/cwgoes/working/go/src/github.com/cosmos/cosmos-sdk/x/mock/simulation/random_simulate_blocks.go:119 +0xb6
panic(0xe53260, 0xc003925900)
	/usr/lib/go/src/runtime/panic.go:513 +0x1b9
github.com/cosmos/cosmos-sdk/x/distribution/keeper.Keeper.WithdrawDelegationRewardsAll(0x1100880, 0xc0001e6ef0, 0xc000a664d0, 0xc000a664d0, 0x1100880, 0xc0001e6f40, 0x11008c0, 0xc0001e6f50, 0xc000b3c600, 0x5, ...)
	/home/cwgoes/working/go/src/github.com/cosmos/cosmos-sdk/x/distribution/keeper/delegation.go:117 +0x512
github.com/cosmos/cosmos-sdk/x/distribution.handleMsgWithdrawDelegatorRewardsAll(0x1107fe0, 0xc00316dad0, 0xc004c09e40, 0xc, 0xc000a7db40, 0x14, 0x20, 0x1100880, 0xc0001e6ef0, 0xc000a664d0, ...)
	/home/cwgoes/working/go/src/github.com/cosmos/cosmos-sdk/x/distribution/handler.go:48 +0xf5
github.com/cosmos/cosmos-sdk/x/distribution.NewHandler.func1(0x1107fe0, 0xc00316dad0, 0xc004c09e40, 0xc, 0x110a7e0, 0xc00316fd60, 0x0, 0x0, 0x0, 0x0, ...)
	/home/cwgoes/working/go/src/github.com/cosmos/cosmos-sdk/x/distribution/handler.go:17 +0x221
github.com/cosmos/cosmos-sdk/x/distribution/simulation.SimulateMsgWithdrawDelegatorRewardsAll.func1(0xc000b458c0, 0xc000b40e10, 0x1107fe0, 0xc001fad6b0, 0xc004c09e40, 0xb, 0xc000a9f000, 0xfa, 0xfa, 0xc0001aa000, ...)
	/home/cwgoes/working/go/src/github.com/cosmos/cosmos-sdk/x/distribution/simulation/msgs.go:58 +0x23c
github.com/cosmos/cosmos-sdk/x/mock/simulation.createBlockSimulator.func2(0x7a, 0xc000b458c0, 0xc000b40e10, 0x1107fe0, 0xc001fad6b0, 0xc004c09e40, 0xb, 0xc000a9f000, 0xfa, 0xfa, ...)
	/home/cwgoes/working/go/src/github.com/cosmos/cosmos-sdk/x/mock/simulation/random_simulate_blocks.go:217 +0x1c9
github.com/cosmos/cosmos-sdk/x/mock/simulation.SimulateFromSeed(0x1112900, 0xc000493100, 0xc000b40e10, 0x102b9e0, 0x2a, 0xc0000d60e0, 0xe, 0xe, 0xc0000b1ef8, 0x0, ...)
	/home/cwgoes/working/go/src/github.com/cosmos/cosmos-sdk/x/mock/simulation/random_simulate_blocks.go:158 +0x10ea
github.com/cosmos/cosmos-sdk/cmd/gaia/app.TestFullGaiaSimulation(0xc000493100)
	/home/cwgoes/working/go/src/github.com/cosmos/cosmos-sdk/cmd/gaia/app/sim_test.go:192 +0x2a9
testing.tRunner(0xc000493100, 0x102b9b0)
	/usr/lib/go/src/testing/testing.go:827 +0xbf
created by testing.(*T).Run
	/usr/lib/go/src/testing/testing.go:878 +0x353

```

cwgoes commented:
Somehow `accum` for a particular validator is ending up greater than the total:

```bash
=== RUN   TestFullGaiaSimulation
Starting SimulateFromSeed with randomness created with seed 42
Starting the simulation from time Sun Jan 31 13:59:47 UTC 26140, unixtime 762734152787
Simulating... block 29/400, operation 350/499.  blocks: 6
accum: 9240000000000
pool: [{steak  39936277489}]
accum: 9240000000000
val accum: 8873500000000
withdrawal: [{steak 41585755790}]
remaining: [{steak -1649478301}]
Panic with err
 negative rem
goroutine 35 [running]:
runtime/debug.Stack(0xc004927410, 0x2, 0x2)
	/usr/lib/go/src/runtime/debug/stack.go:24 +0xa7
```

rigelrozanski commented on x/stake/handler.go:
```diff
@@ -114,8 +114,6 @@ func handleMsgCreateValidator(ctx sdk.Context, msg types.MsgCreateValidator, k k
 
 	k.OnValidatorCreated(ctx, validator.OperatorAddr)
-	accAddr := sdk.AccAddress(validator.OperatorAddr)
-	k.OnDelegationCreated(ctx, accAddr, validator.OperatorAddr)
```
can't delete these, there is a delegation created (self delegation) which needs to be initialized

cwgoes replied:
it's now initialized with the `.Delegate()` call

rigelrozanski replied:
aha - yeah okay got it

rigelrozanski commented on x/stake/handler.go:
```diff
@@ -150,6 +148,8 @@ func handleMsgEditValidator(ctx sdk.Context, msg types.MsgEditValidator, k keepe
 			return err.Result()
 		}
 		validator.Commission = commission
+		// call the hook if present
```
I don't think we need this comment

cwgoes replied:
Removed

rigelrozanski commented on x/stake/keeper/delegation.go:
```diff
@@ -359,6 +359,13 @@ func (k Keeper) Delegate(ctx sdk.Context, delAddr sdk.AccAddress, bondAmt sdk.Co
+		k.OnDelegationSharesModified(ctx, delAddr, validator.OperatorAddr)
+	} else {
+		k.OnDelegationCreated(ctx, delAddr, validator.OperatorAddr)
+	}
```
this looks fine ++

rigelrozanski commented on x/stake/keeper/delegation.go:
```diff
@@ -567,7 +575,6 @@ func (k Keeper) BeginRedelegation(ctx sdk.Context, delAddr sdk.AccAddress,
 	if err != nil {
 		return types.Redelegation{}, err
 	}
-	k.OnDelegationCreated(ctx, delAddr, valDstAddr)
```
k - so this change is fine because `OnDelegationModified` called within delegate for the new delegation. ++

rigelrozanski commented on x/mint/expected_keepers.go:
```diff
@@ -1,9 +1,14 @@
 // expected stake keeper
 type StakeKeeper interface {
+	GetPool(ctx sdk.Context) stake.Pool
+	SetPool(ctx sdk.Context, pool stake.Pool)
```
we should just use nah use the function `InflateSupply` (already here) instead

cwgoes replied:
ah of course, missed that

rigelrozanski commented on x/mint/abci_app.go:
```diff
@@ -21,5 +21,8 @@ func BeginBlocker(ctx sdk.Context, k Keeper) {
 	k.fck.AddCollectedFees(ctx, sdk.Coins{mintedCoin})
+	pool := k.sk.GetPool(ctx)
+	pool.LooseTokens = pool.LooseTokens.Add(sdk.NewDecFromInt(mintedCoin.Amount))
+	k.sk.SetPool(ctx, pool)
```
let's use InflateSupply instead of performing pool operations

cwgoes replied:
yup

cwgoes commented:
Fixed two issue w.r.t minting: the stake pool loose token supply wasn't being updated, and the token amounts used by the simulation weren't enough to have any inflation (it was just truncated to zero). Now we're actually simulating inflation and it seems to work - back to the original error:

```bash
=== RUN   TestFullGaiaSimulation
Starting SimulateFromSeed with randomness created with seed 42
Starting the simulation from time Sun Jan 31 13:59:47 UTC 26140, unixtime 762734152787
Simulating... block 36/400, operation 200/746.  Panic with err
 individual accum should never be greater than the total
goroutine 11 [running]:
runtime/debug.Stack(0xc005965868, 0x2, 0x2)
	/usr/lib/go/src/runtime/debug/stack.go:24 +0xa7
github.com/cosmos/cosmos-sdk/x/mock/simulation.SimulateFromSeed.func3(0xc000c90b60, 0xc000200e00, 0xc000b3cc10)
	/home/cwgoes/working/go/src/github.com/cosmos/cosmos-sdk/x/mock/simulation/random_simulate_blocks.go:119 +0xb6
panic(0xd535c0, 0x10fa280)
	/usr/lib/go/src/runtime/panic.go:513 +0x1b9
github.com/cosmos/cosmos-sdk/x/distribution/types.ValidatorDistInfo.TakeFeePoolRewards(0xc0050e2440, 0x14, 0x14, 0x24, 0x0, 0x0, 0x0, 0xc0050e68a0, 0x1, 0x1, ...)
	/home/cwgoes/working/go/src/github.com/cosmos/cosmos-sdk/x/distribution/types/validator_info.go:49 +0x593
github.com/cosmos/cosmos-sdk/x/distribution/types.DelegationDistInfo.WithdrawRewards(0xc0050e23a0, 0x14, 0x14, 0xc0050e23c0, 0x14, 0x14, 0x1f, 0x24, 0xc0050e6560, 0xc0050e6660, ...)
	/home/cwgoes/working/go/src/github.com/cosmos/cosmos-sdk/x/distribution/types/delegator_info.go:35 +0x2cd
github.com/cosmos/cosmos-sdk/x/distribution/keeper.Keeper.getDelegatorRewardsAll.func1(0x2, 0x11079e0, 0xc001e2ef80, 0xc001e2ef80)
	/home/cwgoes/working/go/src/github.com/cosmos/cosmos-sdk/x/distribution/keeper/delegation.go:136 +0x692
github.com/cosmos/cosmos-sdk/x/stake/keeper.Keeper.IterateDelegations(0x1100900, 0xc000b3c960, 0x1100940, 0xc000b3c970, 0xc0000e27e0, 0x110d7e0, 0xc0001b2f00, 0x0, 0x0, 0xc0000e27e0, ...)
	/home/cwgoes/working/go/src/github.com/cosmos/cosmos-sdk/x/stake/keeper/sdk_types.go:118 +0x330
```

rigelrozanski commented:
One additional thing we should make sure we're addressing - the validatorModified hook should be called right before a slashing takes place

cwgoes commented:
> One additional thing we should make sure we're addressing - the validatorModified hook should be called right before a slashing takes place

Ah, that might not be happening, that could also be a problem.

cwgoes commented:
Interestingly, when I disable slashing I get an even faster simulation failure (same panic on negative accum).

cwgoes commented:
Cleaned up debugging prints, removed a few unnecessary changes, and updated PENDING.md to reflect what this PR changes.

This has passed 400-block single seed sim (locally and CI) & 100-block multi-seed sim (ran locally).

cwgoes commented:
We should add accum invariants, but that can probably happen separately - https://github.com/cosmos/cosmos-sdk/issues/2573.

rigelrozanski commented on x/slashing/keeper.go:
```diff
@@ -37,7 +37,8 @@ func NewKeeper(cdc *codec.Codec, key sdk.StoreKey, vs sdk.ValidatorSet, paramspa
 
-// handle a validator signing two blocks at the same height
+// handle a validator signing two blocks at the same height.
+// power: power of the double-signing validator at the height of infraction.
```
improper use of capitals and periods

cwgoes replied:
Updated.

rigelrozanski commented on x/slashing/keeper.go:
```diff
@@ -70,7 +71,12 @@ func (k Keeper) handleDoubleSign(ctx sdk.Context, addr crypto.Address, infractio
+	// Tendermint.  This value is validator.Tokens as sent to Tendermint via
+	// ABCI, and now received as evidence.
+	// The revisedFraction (which is the new fraction to be slashed) is passed
+	// in separately to separately slash unbonding and rebonding delegations.
```
Improper use of capitals and periods

cwgoes replied:
Removed some periods, not sure which capitals you're referring to.

rigelrozanski commented on x/stake/keeper/keeper.go:
```diff
@@ -71,21 +71,73 @@ func (k Keeper) SetPool(ctx sdk.Context, pool types.Pool) {
+func (k Keeper) powerToBytes(power sdk.Dec) []byte {
+	bz := k.cdc.MustMarshalBinary(power)
+	return bz
+}
```
not sure this function should exist - I think we should just marshal in place

cwgoes replied:
OK, we now marshal in-place.

cwgoes commented on x/distribution/keeper/validator.go:
```diff
@@ -34,27 +41,52 @@ func (k Keeper) RemoveValidatorDistInfo(ctx sdk.Context, valAddr sdk.ValAddress)
+
+	if !k.HasValidatorDistInfo(ctx, operatorAddr) {
+		return types.ErrNoValidatorDistInfo(k.codespace)
+	}
```
https://github.com/cosmos/cosmos-sdk/pull/2501#discussion_r227512145

cwgoes commented on x/distribution/types/delegator_info.go:
```diff
@@ -21,7 +21,13 @@ func NewDelegationDistInfo(delegatorAddr sdk.AccAddress, valOperatorAddr sdk.Val
+// * calls vi.TakeFeePoolRewards, which:
+//   * updates validator info's FeePoolWithdrawalHeight, thus setting accum to 0.
+//   * updates fee pool to latest height and total val accum w/ given totalBonded.
+//   (see comment on TakeFeePoolRewards for more info).
```
OK, updated.

cwgoes commented on x/distribution/types/dec_coin.go:
```diff
@@ -147,3 +152,27 @@ func (coins DecCoins) QuoDec(d sdk.Dec) DecCoins {
+		}
+		return sdk.ZeroDec()
+	default:
+		midIdx := len(coins) / 2 // 2:1, 3:1, 4:2
```
It is, I think it was copied. This is just binary search. Updated.

cwgoes commented on x/distribution/keeper/delegation.go:
```diff
@@ -62,29 +69,40 @@ func (k Keeper) RemoveDelegatorWithdrawAddr(ctx sdk.Context, delAddr, withdrawAd
-// withdraw all the rewards for a single delegation
+// Withdraw all the rewards for a single delegation.
+// NOTE: This gets called "onDelegationSharesModified",
+// meaning any changes to bonded coins.
```
Updated.

cwgoes commented on types/decimal.go:
```diff
@@ -247,6 +247,11 @@ func (d Dec) QuoInt(i Int) Dec {
 	return Dec{mul}
 }
 
+// is integer, e.g. decimals are zero.
```
Updated.

cwgoes commented on x/distribution/types/validator_info.go:
```diff
@@ -31,27 +31,39 @@ func (vi ValidatorDistInfo) UpdateTotalDelAccum(height int64, totalDelShares sdk
+// - called in ValidationDistInfo.WithdrawCommission.
+// - called in DelegationDistInfo.WithdrawRewards.
+// NOTE: When a delegator unbonds, say, onDelegationSharesModified ->
+//       WithdrawDelegationReward -> WithdrawRewards.
```
Updated, removed periods on sentence fragments.

rigelrozanski commented on x/distribution/keeper/delegation.go:
```diff
@@ -62,29 +69,40 @@ func (k Keeper) RemoveDelegatorWithdrawAddr(ctx sdk.Context, delAddr, withdrawAd
 
+	if !k.HasDelegationDistInfo(ctx, delegatorAddr, valAddr) {
+		return types.ErrNoDelegationDistInfo(k.codespace)
+	}
```
OH my mistake I was confusing this with the staking hooks

rigelrozanski commented on x/stake/keeper/keeper.go:
```diff
@@ -71,21 +71,73 @@ func (k Keeper) SetPool(ctx sdk.Context, pool types.Pool) {
+func (k Keeper) SetLastTotalPower(ctx sdk.Context, power sdk.Dec) {
+	if !power.IsInteger() {
+		panic("input power must be whole integer")
+	}
```
if what you're saying is correct shouldn't this be panicing sometimes in the simulation then?

rigelrozanski replied:
Should change the saved field to just being the integer here likely

jaekwon commented on x/stake/handler.go:
```diff
@@ -150,6 +148,7 @@ func handleMsgEditValidator(ctx sdk.Context, msg types.MsgEditValidator, k keepe
 			return err.Result()
 		}
 		validator.Commission = commission
+		k.OnValidatorModified(ctx, msg.ValidatorAddr)
```
probably shouldn't be calling this before k.SetValidator().  Also hooks being called from handlers is weird.

cwgoes replied:
I think this is the correct order, but we should change the name of the hook - https://github.com/cosmos/cosmos-sdk/issues/2538.
